### PR TITLE
Add lease renewal integration and lease-loss abort to scan loop

### PR DESCRIPTION
**Pull request by ahrav**
## Summary

- Adds caller-configurable lease renewal via `RenewalPolicy` struct (renew-at-fraction with clamping and non-finite fallback)
- Introduces `ScanLoopOutcome::LeaseLost` variant with `LeaseLossCause` diagnostic enum for clean abort-without-mutation semantics
- Splits entry point into `run_scan_loop` (default half-life policy) and `run_scan_loop_with_policy` (configurable)
- Extracts 876-line test module into dedicated `scan_loop_test.rs` file

### Lease renewal logic (`scan_loop.rs`)

| Change | Description |
|--------|-------------|
| `RenewalPolicy` struct | Configurable renew-at-fraction with clamping and non-finite fallback |
| `LeaseLossCause` enum | `RenewFailed(RenewError)` or `DeadlineElapsed { now, deadline }` |
| `ScanLoopOutcome::LeaseLost` | Fourth terminal outcome — abort without issuing `park`/`complete` |
| Pre-checkpoint deadline guard | Checks `now >= deadline` before `checkpoint()` to surface lease loss early |
| Post-checkpoint renewal | Evaluates policy after each successful checkpoint; recalibrates from renewed deadline |
| `should_renew()` helper | Pure function: `(now, deadline, observed_duration, policy) -> bool` |
| `pages_completed` counter | Incremented only on successful checkpoint |

### Lease-loss boundary

`LeaseLost` is intentionally narrow:
- **Before checkpoint** → deadline elapsed → `LeaseLost(DeadlineElapsed)`
- **After checkpoint** → renewal fails → `LeaseLost(RenewFailed)`
- **Mutation attempted** → backend rejects (including lease-expired) → `Error` (not `LeaseLost`)

This gives callers a clear split: "reacquire and resume" vs "coordination mutation failed."

### Public API additions (`lib.rs`)

New exports: `DEFAULT_RENEW_AT_FRACTION`, `LeaseLossCause`, `RenewalPolicy`, `run_scan_loop_with_policy`

### Test file extraction

880-line `mod tests` block moved to `scan_loop_test.rs` via `#[path]` directive. No test logic changed.

## Why

The scan loop had no lease awareness — it would blindly attempt checkpoints on expired leases, getting `LeaseExpired` errors indistinguishable from backend failures. This adds proactive deadline detection, synchronous renewal (DST-compatible), and a clear lease-loss/mutation-failure split.

Required by lease lifecycle design (`diagrams/07-lease-lifecycle.md`). Unblocks gossip-9u1j.15 (tracing + worker binary wiring).

## Test plan

- [x] 3 new lease-specific tests: deadline elapsed, successful renewal, `StaleFence` failure
- [x] All existing scan loop tests pass unchanged after file extraction
- [x] `cargo test -p gossip-scan-pipeline`
- [x] `cargo clippy --all-targets --all-features -- -D warnings`
- [x] `cargo fmt --all`

Closes gossip-9u1j.14 (B4-C6-T2: Lease renewal integration and lease-loss abort)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/gossip-scan-pipeline/src/scan_loop.rs [hunk 1]
```diff
@@ -296,7 +402,16 @@ where
...
     };
 
+    // Seed the renew threshold from the lease time still remaining when the
+    // loop actually starts, not from the run's configured lease duration.
+    let baseline_now = now_fn();
+    let mut observed_lease_duration = session
+        .lease()
+        .deadline()
+        .as_raw()
+        .saturating_sub(baseline_now.as_raw());
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-scan-pipeline/src/scan_loop.rs]: I investigated this and wrote a test targeting the late-entry scenario you described (worker enters with only 10 of 20 ticks remaining). The test passes — the code correctly renews at the proportional threshold (5 ticks remaining at 50% fraction), then recalibrates `observed_lease_duration` to the full lease duration after the first successful renewal.

This behavior is intentional per the doc comment at line 333: *"keeps renewal timing stable even if time has already elapsed between acquire and entering the loop."* The key insight is that after the first renewal, `observed_lease_duration` recalibrates to the full fresh lease window, so subsequent renewals use the full duration.

File: crates/gossip-scan-pipeline/src/scan_loop.rs [hunk 2]
```diff
@@ -372,17 +487,90 @@ where
...
+    let remaining = deadline.as_raw().saturating_sub(now.as_raw());
+    let fraction = if policy.renew_at_fraction.is_finite() {
+        policy.renew_at_fraction.clamp(0.0, 1.0)
+    } else {
+        DEFAULT_RENEW_AT_FRACTION
+    };
+    let mut renew_threshold = ((observed_lease_duration as f64) * fraction).ceil() as u64;
+    if renew_threshold == 0 && fraction > 0.0 {
+        renew_threshold = 1;
+    }
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-scan-pipeline/src/scan_loop.rs]: Confirmed — the edge cases are intentional. Added a dedicated edge-case test (`should_renew_small_duration_edge_cases`) verifying behavior for durations 0, 1, and 2 with various fractions. The `ceil()` rounding is conservative (triggers renewal earlier), and the `threshold == 0 && fraction > 0` guard correctly forces a minimum threshold of 1 tick so that a non-zero renewal fraction always triggers at least one tick before the deadline.

**Comment by ahrav** [comment]
Re: inconsistent deadline handling between `complete` and `checkpoint` paths —

This asymmetry is intentional, as documented in the module-level doc comment (line 75-78) and the function doc (line 340-344):

> `LeaseLost` is intentionally narrow: it is returned only when the loop can detect lease loss before issuing a coordination mutation (`checkpoint`) or when `renew` itself fails. If `checkpoint`, `complete`, or `park` is attempted and rejected by the backend (including lease-expired cases), the outcome remains `Error`.

The checkpoint guard avoids wasting a network round trip on every non-terminal page (hot path). The `complete` call only runs once per shard (terminal page), so the optimization isn't needed there — backend rejection provides the same information. The existing test `complete_failure_returns_error` validates this explicitly: an expired-lease `complete` returns `Error(Complete(LeaseExpired))`, not `LeaseLost`.

Adding a guard before `complete` would change the outcome type from `Error(Complete(LeaseExpired))` to `LeaseLost(DeadlineElapsed)`, which would break the documented semantic boundary between "detected before mutation" and "mutation attempted and rejected."
